### ISSUE TITLE
Training for glasses recognition

### ISSUE THREAD

**Issue by Sexpistrolls** [AlexeyAB/darknet#453]
Hello, I've made a training with people wearing glasses and I let it run all night, when I checked it up I saw that it recognized glasses when he detected eyes ... thing that I don't really want. What do you think I could do ?
I thougth about training it to recognize glasses (not worn by people), or mix both (take last training and train it again with glasses without people), what would I rather do ? Have you got any other idea ? :/ thanks !

**Comment by AlexeyAB** [comment]
Hi, probably in your training dataset there are images that contain people with glasses through which the eyes are visible.
* You can add images to your training dataset with eyes without glasses, and without labels. Just .jpg-files and empty .txt-files, and add paths to these jpg files to the train.txt. So neural network will be trained to detect glasses as object, and eye as not object (background).
* Or you can  training it to recognize glasses (not worn by people).

I recommend to train from the begining.

**Comment by Sexpistrolls** [comment]
Thank you very much !

**Comment by Sexpistrolls** [comment]
Ok, I tryed both of the solutions and I trained from the begining but there's still the same problem and the tests are even crazier than before ... I am quite lost about how the training works in general ... Could you maybe show me an example of one of your trainings, just to see what it looks like (I mean like a screen of a little sample of 10-20 pictures with the boxes and tag to see how you did and what you did want to detect) it would be perfect.
-----------------------------------------------
Here is what I wanted to do with my training :  
(I took about 350 pictures all in all choosed myself from various angles and I let it train during more than 15 hours (both each or all together) and tested every 1000 iterations weigths from 3000 to 22500)
- I have to recognize safety headgear for dangerous sites
- safety glasses/googles on people
- recognize both standing and lying/fainted people 
- orange and green fluo jackets (I need to make the difference between the two colours)
-------------------------
The detection issues : 
- Detect glasses even if none are worn (I put normal, big and sunprotect ones and from different angles, worn by people or not)
- No real problem to tell the difference between colour jackets but sometimes detect blue helmet/headgear like an orange jacket
- No problem to detect standing people but hesitate far too much when people are lying on ground, sometimes double the detection and put both of the tags 

Sorry for this long post but I'm quite desparate and I hope I'll see what to do with some examples or explaination (and thank you for your quick answers, your incredible hahah).

**Comment by AlexeyAB** [comment]
The simplest dataset for training (11 images for 2 classes) is here: [dataset.zip](https://github.com/AlexeyAB/darknet/files/1812966/dataset.zip)

The same images you can find here - this is the utility that you should use for label your images (Linux or Windows): https://github.com/AlexeyAB/Yolo_mark

* Did you use this utility to label your images? https://github.com/AlexeyAB/Yolo_mark
* What mAP, IoU, TP/FP/FN can use get using `./darknet detector map obj.data obj.cfg obj.weights`
* Can you show some wrong detected images from your dataset or from internet?
* If you want to distinguish colors, you should change these params in the cfg-file `saturation = 1.2 exposure = 1.5 hue=.02`: https://github.com/AlexeyAB/darknet/blob/df076653e00db69c6fb57869981f5196a8f55e70/cfg/yolo-voc.2.0.cfg#L10-L12

----

![image](https://user-images.githubusercontent.com/4096485/37429501-8a4ac83e-27e0-11e8-8854-908ef0775939.png)

**Comment by CruisWoo** [comment]
@AlexeyAB Hello!I used Yolo v2 to train my own data(1 class) follw your instructions,after training,I want to get mAP, IoU, TP/FP/FN,etc.But when I use ./darknet detector map data/obj.data yolo-obj.cfg yolo-obj_final.weights ,the terminal didn't show anything.
When I use ./darknet detector recall data/obj.data yolo-obj.cfg yolo-obj_final.weights,it can show like below:
7585 7611 7688 RPs/Img: 68.22 IOU: 77.86% Recall:99.00%
7586 7612 7689 RPs/Img: 68.23 IOU: 77.86% Recall:99.00%
......
but no map,no precision,
so,how can I calculate the AP(I remember 1class mAP should be AP)and precision?
Should I change code?Or using other command?

**Comment by AlexeyAB** [comment]
You can also download this fork: https://github.com/AlexeyAB/darknet
And compile it. I implemented ` ./darknet detector map `...

**Comment by Sexpistrolls** [comment]
I already was using Yolo_mark. At the begining I was training for only 2 or 3 tags and all was clear and efficient like for the two coloured jackets and my very first trainings were with the basic datasets that were given like, with coco and so on, so I can tell that it was working at the begining. I'll try to change the settings for the colours that you recommand and test it tomorrow and I'll send you several results then. But anyway I think it'll always be too difficult to recognize lying people from standing ones because of the angles, wall, etc. From what I saw I would tell that a standing person against a wall is far to similar from a person lying on the ground (flat surface and person in a straight position in each case even if I personnally take pictures from every possible angle) , I don't know what you think about that.

**Comment by Sexpistrolls** [comment]
Yeah I saw it and thought about that but there's then a probleme when the person is lying "facing you", from a camera looking at it from above it just seems vertical, I don't know if you see what I mean : 
http://laoblogger.com/images/people-lying-on-floor-clipart-8.jpg

**Comment by AlexeyAB** [comment]
Here there are details that indicate that this person is lying, for example, the soles are visible, and the chin is visible instead of the face.
Each final activation of Yolo v2 has a receptive field of 566x566 pixels, so even with 544x544 network resolution, some final activations will see the whole person in context, so network will be able to distinguish the lying person from the standing one.

**Comment by Sexpistrolls** [comment]
Ok now I see, thank you very much ! 
But I don't really know what to do except keeping trying and make a better dataset. 
I guess I'll try to train with more pictures I guess than (I don't know if I told you but when I tryed, it was boxing with both of the standing and lying tags).
How many picture do you recommand me to have for it and for the whole 6-tags-training ? And have you got some recommandations or details about how to make it more properly ?

**Comment by AlexeyAB** [comment]
At least 2000 images per class. So if you use 6 classes, you should have at least 12 000 labeled images. More the better.

How to improve object detection: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

**Comment by Sexpistrolls** [comment]
Oh ... that's a lot ... , Can I ask you where do you find your pictures for your training ? (and did mark each one by hand with yolo_mark ?).
It took me a lot of time for only 300 so I don't even wonder for 12000, I tryed to use image free bank Image.net but I couldn't find enough or effecient pictures, so if you have some advice or if you know a great place where to easily get hundreds of it I'll take it.
And finally, thank you very much for your time and all your answers.

**Comment by AlexeyAB** [comment]
Yes, it takes a lot of time. But if your task is not critical, then you can take less pictures (100 per class).

* You can get frames from some video, and manually do labels: https://github.com/AlexeyAB/Yolo_mark#how-to-get-frames-from-videofile
* Or you can use labled images from VOC, COCO or ImageNet datasets.

**Comment by Sexpistrolls** [comment]
Hello again, I was wondering how the training detects thing that are not boxed in the pictures because most of the needed pictures that I found are marked by copyrights or writings like "Shuttershock" if you know it : 
https://www.shutterstock.com/fr/image-photo/sao-paulo-brazil-circa-jan-14-178075751

----
I don't know if it's a problem or not, do I have to pick only unmarked pictures ? (I guess it would be better)

**Comment by AlexeyAB** [comment]
Yes, much better to use photos without water marks.

**Comment by Sexpistrolls** [comment]
And now, I'm making a setup with a group of about 30 students to take bunches of photos from a scene/place,  I'll make them move, lean/stand, where jackets, face the camera or not and so on, I guess it will do the trick.
